File: app/src/main/java/com/spiderbiggen/randomchampionselector/model/Champion.java
package com.spiderbiggen.randomchampionselector.model;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Embedded;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Locale;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;


/**
 * Class that defines the champion object.
 *
 * @author Stefan Breetveld
 */
@Entity
public class Champion implements Serializable, Parseable<Champion> {

    @PrimaryKey
    private int key;
    private String id;
    private String name;
    private String title;
    private String lore;
    private String blurb;
    @ColumnInfo(name = "roles")
    private String[] tags;
    @Embedded
    private Info info;
    @Embedded
    private Image image;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public int getKey() {
        return key;
    }

    public void setKey(int key) {
        this.key = key;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    @Ignore
    public String getCapitalizedTitle() {
        return title.substring(0, 1).toUpperCase(Locale.ENGLISH) + title.substring(1);
    }

    public String getLore() {
        return lore;
    }

    public void setLore(String lore) {
        this.lore = lore;
    }

    public String getBlurb() {
        return blurb;
    }

    public void setBlurb(String blurb) {
        this.blurb = blurb;
    }

    public String[] getTags() {
        return tags;
    }

    public void setTags(String[] tags) {
        this.tags = tags;
    }

    public Info getInfo() {
        return info;
    }

    public void setInfo(Info info) {
        this.info = info;
    }

    public Image getImage() {
        return image;
    }

    public void setImage(Image image) {
        this.image = image;
    }

    /**
     * Get this champion's attack.
     *
     * @return null if info or the result of {@link Info#getAttack()}
     * @see Info#getAttack()
     */
    public Byte getAttack() {
        if (info == null) {
            return null;
        }
        return info.getAttack();
    }

    public Byte getDefense() {
        if (info == null) {
            return null;
        }
        return info.defense;
    }

    public Byte getMagic() {
        if (info == null) {
            return null;
        }
        return info.getMagic();
    }

    public Byte getDifficulty() {
        if (info == null) {
            return null;
        }
        return info.getDifficulty();
    }

    public String getFull() {
        if (image == null) {
            return null;
        }
        return image.getFull();
    }

    public String getSprite() {
        if (image == null) {
            return null;
        }
        return image.getSprite();
    }

    public String getGroup() {
        if (image == null) {
            return null;
        }
        return image.getGroup();
    }

    public Integer getX() {
        if (image == null) {
            return null;
        }
        return image.getX();
    }

    public Integer getY() {
        if (image == null) {
            return null;
        }
        return image.getY();
    }

    public Integer getW() {
        if (image == null) {
            return null;
        }
        return image.getW();
    }

    public Integer getH() {
        if (image == null) {
            return null;
        }
        return image.getH();
    }

    @Override
    public String toString() {
        return "Champion{" +
            "key=" + key +
            ", name='" + name + '\'' +
            ", roles='" + Arrays.toString(tags) + '\'' +
            '}';
    }

    @Override
    public Champion parse(JSONObject object) throws JSONException {
        setKey(object.getInt("key"));
        setId(object.getString("id"));
        setBlurb(object.optString("blurb"));
        setLore(object.optString("lore"));
        setTitle(object.optString("title"));
        setName(object.optString("name"));
        JSONArray array = object.optJSONArray("tags");
        if (array != null) {
            String[] tags = new String[array.length()];
            for (int i = 0; i < array.length(); i++) {
                tags[i] = array.getString(i);
            }
            setTags(tags);
        }
        JSONObject infoObject = object.getJSONObject("info");
        if (infoObject != null) {
            Info info = new Info();
            setInfo(info.parse(infoObject));
        }
        JSONObject imageObject = object.getJSONObject("image");
        if (imageObject != null) {
            Image image = new Image();
            setImage(image.parse(imageObject));
        }
        return this;
    }

    public static class Info implements Serializable, Parseable<Info> {
        private byte attack;
        private byte defense;
        private byte magic;
        private byte difficulty;

        public byte getAttack() {
            return attack;
        }

        public void setAttack(byte attack) {
            this.attack = attack;
        }

        public byte getDefense() {
            return defense;
        }

        public void setDefense(byte defense) {
            this.defense = defense;
        }

        public byte getMagic() {
            return magic;
        }

        public void setMagic(byte magic) {
            this.magic = magic;
        }

        public byte getDifficulty() {
            return difficulty;
        }

        public void setDifficulty(byte difficulty) {
            this.difficulty = difficulty;
        }

        @Override
        public Info parse(JSONObject object) throws JSONException {
            setAttack((byte) object.optInt("attack", 0));
            setDefense((byte) object.optInt("defense", 0));
            setMagic((byte) object.optInt("magic", 0));
            setDifficulty((byte) object.optInt("difficulty", 0));
            return this;
        }
    }

    public static class Image implements Serializable, Parseable<Image> {
        private String full;
        private String sprite;
        private String group;
        private int x;
        private int y;
        private int w;
        private int h;

        public String getFull() {
            return full;
        }

        public void setFull(String full) {
            this.full = full;
        }

        public String getSprite() {
            return sprite;
        }

        public void setSprite(String sprite) {
            this.sprite = sprite;
        }

        public String getGroup() {
            return group;
        }

        public void setGroup(String group) {
            this.group = group;
        }

        public int getX() {
            return x;
        }

        public void setX(int x) {
            this.x = x;
        }

        public int getY() {
            return y;
        }

        public void setY(int y) {
            this.y = y;
        }

        public int getW() {
            return w;
        }

        public void setW(int w) {
            this.w = w;
        }

        public int getH() {
            return h;
        }

        public void setH(int h) {
            this.h = h;
        }

        @Override
        public Image parse(JSONObject object) throws JSONException {
            setFull(object.getString("full"));
            setSprite(object.getString("sprite"));
            setGroup(object.getString("group"));
            setX(object.optInt("x", 0));
            setY(object.optInt("y", 0));
            setW(object.optInt("w", 0));
            setH(object.optInt("x", 0));
            return this;
        }
    }
}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/domain/Champion.kt
package com.spiderbiggen.randomchampionselector.domain

import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.PrimaryKey
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.Serializable
import java.util.*

/**
 * Class that defines the champion object.
 *
 * @author Stefan Breetveld
 */
@Entity
data class Champion(
        @PrimaryKey val key: Int,
        val id: String,
        val name: String,
        val title: String,
        val lore: String,
        val blurb: String,
        val roles: Array<String>,
        @Embedded val info: Info
) : Serializable {

    val capitalizedTitle: String
        get() = title.substring(0, 1).toUpperCase(Locale.ENGLISH) + title.substring(1)

    override fun toString(): String {
        return "Champion(key=$key, name='$name', roles=${Arrays.toString(roles)})"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Champion) return false

        return key == other.key && id == other.id && name == other.name
    }

    override fun hashCode(): Int {
        var result = key
        result = 31 * result + id.hashCode()
        result = 31 * result + name.hashCode()
        return result
    }

    companion object : Parsable<Champion> {
        override fun parse(jsonObject: JSONObject): Champion {
            val key = jsonObject.getInt("key")
            val id = jsonObject.getString("id")
            val blurb = jsonObject.optString("blurb")
            val lore = jsonObject.optString("lore")
            val title = jsonObject.optString("title")
            val name = jsonObject.optString("name")
            val roles = jsonArrayToStringArray(jsonObject.optJSONArray("tags"))
            val info = Info.parse(jsonObject.getJSONObject("info"))
            return Champion(key, id, name, title, lore, blurb, roles, info)
        }

        private fun jsonArrayToStringArray(jsonArray: JSONArray): Array<String> {
            val tags = (0 until jsonArray.length()).map(jsonArray::getString)
            return tags.toTypedArray()
        }
    }

    data class Info(val attack: Byte, val defense: Byte, val magic: Byte, val difficulty: Byte) : Serializable {
        companion object : Parsable<Info> {
            @Throws(JSONException::class)
            override fun parse(jsonObject: JSONObject): Info {
                val attack = jsonObject.optInt("attack", 0).toByte()
                val defense = jsonObject.optInt("defense", 0).toByte()
                val magic = jsonObject.optInt("magic", 0).toByte()
                val difficulty = jsonObject.optInt("difficulty", 0).toByte()
                return Info(attack, defense, magic, difficulty)
            }
        }

    }
}
